File: fl_backend/app/api/kontakte/schemas.py
from typing import Annotated

from pydantic import BaseModel, ConfigDict, EmailStr, Field, StringConstraints

# The wire's spelling of the three seats, imported rather than restated: a second closed set here
# would let this read name a seat no other endpoint publishes.
from app.api.bewerbungen.schemas import FLKontaktRolle
from app.shared.schemas.bounds import KONTAKT_EMAIL_MAX_LENGTH
from app.shared.schemas.responses import BaseAPIResponse


class FLKontaktErasurePayload(BaseModel):
    """Which person is asking to be forgotten, named by the one field that joins their records.

    A contact block carries no id and nothing joins one season's Trainer to the next season's, so
    the address IS the identity here.
    """

    model_config = ConfigDict(extra="forbid")

    # In the BODY and on no path or query: an address in a path lands in the access log, in nginx's
    # log and in `aktionen.request.path`, three fresh copies of the value the request exists to destroy.
    email: Annotated[EmailStr, StringConstraints(max_length=KONTAKT_EMAIL_MAX_LENGTH)]


class FLKontaktSitz(BaseModel):
    """One seat the address holds.

    Per seat and not per person: colleagues share a school inbox, and one row seats one person twice
    where `trainer_ist_zugleich` says so, so a list keyed on the name hides both cases.
    """

    saison_id: str
    # Beside the name because two seats of one season are otherwise two rows a reader cannot tell
    # apart, which is exactly the shape `trainer_ist_zugleich` produces.
    rolle: FLKontaktRolle
    vorname: str
    nachname: str


class FLKontaktErasureAnsichtResponse(BaseAPIResponse):
    """Whom `POST /kontakte/erasure` would reach, and nothing it would destroy.

    The two lists are the two collections the write clears, kept apart because an application is a
    request to join and a junction row is a season already played.
    """

    saison_teams: list[FLKontaktSitz]
    bewerbungen: list[FLKontaktSitz]


class FLKontaktErasureResponse(BaseAPIResponse):
    """What the erasure reached, and NOT an echo of the person.

    `app/api/spieler/schemas.py :: FLSpielerErasureResponse`'s reason, binding harder here: the
    request named an address, so echoing anything of theirs returns a copy of what was destroyed.
    """

    # Junction rows and applications whose block named them. Rows, not people: one row can hold the
    # same person twice.
    cleared_saison_teams: int = Field(ge=0)
    cleared_bewerbungen: int = Field(ge=0)
    # The slots actually nulled, across both collections. Higher than the two counts above ADDED
    # wherever `trainer_ist_zugleich` put one person in two slots of one row, which is what
    # makes this the figure showing the double-slot case was reached.
    cleared_kontakt_slots: int = Field(ge=0)
    # Log rows emptied and stamped (`docs/backend/spec.md :: I42`): those naming a row above, and
    # those still HOLDING the person where a swap moved them out of one. Disjoint, so nothing is
    # counted twice -- and no row is dropped, so never a deletion count.
    redacted_aktionen: int = Field(ge=0)

File: fl_frontend/src/features/kontakte/schemas.ts
import z from "zod";

import { BaseAPIResponseSchema } from "@/core/schemas";
import { FLSaisonTeamKontaktePayloadSchema, FLSaisonTeamKontakteSchema } from "@/features/teams/schemas";
import { CustomObjectIdStringSchema, KONTAKT_EMAIL_MAX_LENGTH } from "@/shared/schemas";

/**
 * The address IS the identity: nothing joins one season's Trainer to the next, so the request names
 * a person and not a row. Its own declaration, so no value typed for a write reaches the deletion.
 */
export const FLKontaktErasurePayloadSchema = z.object({
  // Spelled as the team editor spells it: the backend types the field the same way, and its refusal
  // carries no field detail, so an address only the server judged would mark no box.
  email: z
    .email({ error: "Bitte gib eine gültige E-Mail-Adresse ein." })
    .max(KONTAKT_EMAIL_MAX_LENGTH, { error: `Die E-Mail-Adresse darf höchstens ${String(KONTAKT_EMAIL_MAX_LENGTH)} Zeichen lang sein.` }),
});
export type FLKontaktErasurePayload = z.infer<typeof FLKontaktErasurePayloadSchema>;

/**
 * Mirrors `FLKontaktErasureResponse` — counts alone, and deliberately no echo of the person. The
 * request named an address, so answering with anything of theirs would hand back a fresh copy of
 * exactly what the request destroyed.
 */
export const FLKontaktErasureResponseSchema = BaseAPIResponseSchema.extend({
  cleared_saison_teams: z.int().nonnegative(),
  cleared_bewerbungen: z.int().nonnegative(),
  /**
   * The slots actually nulled, across both collections. Higher than the two row counts wherever
   * `trainer_ist_zugleich` seated one person twice in one row, which is what makes this the
   * figure the report counts contact entries by.
   */
  cleared_kontakt_slots: z.int().nonnegative(),
  /** Log rows whose image was emptied and stamped. Never a deletion count: no row is dropped. */
  redacted_aktionen: z.int().nonnegative(),
});
export type FLKontaktErasureResponse = z.infer<typeof FLKontaktErasureResponseSchema>;

/**
 * Mirrors `FLPatchSaisonTeamKontaktePayload` — the three seats alone, on the row the path names.
 * `FLSaisonTeamKontaktePayloadSchema` is reused: the junction PATCH takes the same block, and a
 * second spelling would drift with nothing able to see it.
 */
export const FLPatchSaisonTeamKontaktePayloadSchema = z.object({
  // Both ids are in the PATH on the wire — the junction row is addressed by its natural key. They
  // are carried here because the form has to know which club's season it is writing.
  team_id: CustomObjectIdStringSchema,
  saison_id: z.string().length(4, { error: "Die Saison-ID besteht aus genau 4 Zeichen." }),
  // The whole block, or `null` to clear it. REQUIRED with no default: a form that omits it gets a
  // 422, never three people quietly left standing.
  kontakte: FLSaisonTeamKontaktePayloadSchema.nullable(),
});
export type FLPatchSaisonTeamKontaktePayload = z.infer<typeof FLPatchSaisonTeamKontaktePayloadSchema>;

/**
 * Mirrors `FLPatchSaisonTeamKontakteResponse` — the block as stored after the write, and no other
 * field of the row. The endpoint answers about the seats it moved, so the group and the Austritt
 * beside them are not its to echo.
 */
export const FLPatchSaisonTeamKontakteResponseSchema = BaseAPIResponseSchema.extend({
  saison_id: z.string(),
  team_id: CustomObjectIdStringSchema,
  kontakte: FLSaisonTeamKontakteSchema.nullable(),
});
export type FLPatchSaisonTeamKontakteResponse = z.infer<typeof FLPatchSaisonTeamKontakteResponseSchema>;
